### PR TITLE
fix!: Make fields on S3 structs mandatory, add S3 helpers

### PR DESCRIPTION
**Pull request by sbernauer**
# Description

Reviving https://github.com/stackabletech/operator-rs/pull/679 (replacing it)

Example usage is in https://github.com/stackabletech/trino-operator/tree/chore/new-s3-structs

## Definition of Done Checklist

- Not all of these items are applicable to all PRs, the author should update this template to only leave the boxes in that are relevant
- Please make sure all these things are done and tick the boxes

```[tasklist]
# Author
- [ ] Changes are OpenShift compatible
- [ ] CRD changes approved
- [ ] Integration tests passed (for non trivial changes)
```

```[tasklist]
# Reviewer
- [ ] Code contains useful comments
- [ ] (Integration-)Test cases added
- [ ] Documentation added or updated
- [ ] Changelog updated
- [ ] Cargo.toml only contains references to git tags (not specific commits or branches)
```

```[tasklist]
# Acceptance
- [ ] Feature Tracker has been updated
- [ ] Proper release label has been added
```

### CODE REVIEW

**Review by NickLarsenNZ — COMMENTED** [review]
Partial review (as I see changes are still coming in)

File: crates/stackable-operator/src/commons/s3/helpers.rs
Comment (NickLarsenNZ) [inline crates/stackable-operator/src/commons/s3/helpers.rs:106]: This particular change is breaking, right?

Once ready for review, I think there should be some text in a comment ready for adding to release notes (and the `release-note/action-required` label.

**Review by NickLarsenNZ — COMMENTED** [review]
Partial review again :)

File: crates/stackable-operator/src/commons/s3/mod.rs
Comment (sbernauer) [reply crates/stackable-operator/src/commons/s3/mod.rs]: Only `RetrieveS3Bucket` is unused, removed in 687141c

Comment (sbernauer) [reply crates/stackable-operator/src/commons/s3/mod.rs]: Added some error details in 72417ff

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [inline crates/stackable-operator/src/commons/networking.rs:58]: note: Currently, our naming is a little wonky. RFC 1123 states that:

> Whenever a user inputs the identity of an Internet host, it SHOULD
> be possible to enter either (1) a host domain name or (2) an IP
> address in dotted-decimal ("#.#.#.#") form.

So a host name is either a valid domain name, which is also enforced by this sentence in the RFC:

> Host software MUST handle host names of up to 63 characters and
> SHOULD handle host names of up to 255 characters.

or an IP address. Therefore, the enum should be defined as:

```suggestion
pub enum Hostname {
    IpAddress(IpAddr),
    Domain(DomainName),
}
```

---

A nice thing we already do correctly is:

> The host SHOULD check the string syntactically for a dotted-decimal
> number before looking it up in the Domain Name System.

Comment (Techassi) [reply crates/stackable-operator/src/commons/networking.rs:58]: note: Also on that note, we might want to take a look at the current validation and ensure all regular expressions are nailed down according to the RFCs.

But that is unrelated to this PR, just a general thought I had.

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [reply crates/stackable-operator/src/commons/networking.rs:98]: note: I'm personally a fan of writing custom `Deserialize` and `Serialize` implementations, but I wouldn't block this PR on it.

**Review by Techassi — CHANGES_REQUESTED** [review]
I didn't look at the S3 stuff super closely (I can if you want @sbernauer). The TLS code looks fine (as it is copy+pasted from the previous location).

All other comments are minor notes / suggestions.

File: crates/stackable-operator/src/validation.rs
Comment (Techassi) [reply crates/stackable-operator/src/validation.rs:97]: Also, we should maybe include the input which failed to parse in the error.

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [reply crates/stackable-operator/src/commons/networking.rs:82]: I would really like to see us using `snafu`'s context selector here, but I guess this is a limitation of bundling all validation errors into a single (foreign) enum.

File: crates/stackable-operator/src/validation.rs
Comment (sbernauer) [reply crates/stackable-operator/src/validation.rs]: Added the input in 4fb9d50

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [inline crates/stackable-operator/src/commons/networking.rs:56]: note: Error fields of type `String` should use the debug output, which automatically adds double quotes around it. See [guide](https://docs.stackable.tech/home/stable/contributor/code-style-guide#_error_messages)

```suggestion
        "the given hostname {hostname:?} is not a valid hostname, which needs to be either a domain name or IP address"
```

File: crates/stackable-operator/CHANGELOG.md
Comment (Techassi) [inline crates/stackable-operator/CHANGELOG.md:13]: suggestion: Slight rewording.

```suggestion
- BREAKING: The TLS verification struct now resides in the `commons::tls_verification` module, instead of being placed below `commons::authentication::tls` ([#863]).
```

File: crates/stackable-operator/CHANGELOG.md
Comment (Techassi) [inline crates/stackable-operator/CHANGELOG.md:18]: suggestion: Slight rewording.

```suggestion
- BREAKING: The fields `bucketName`, `connection` and `host` on `S3BucketSpec`, `InlinedS3BucketSpec` and `S3ConnectionSpec` are now mandatory. Previously operators errored out in case these fields where missing ([#863]).
```

File: crates/stackable-operator/CHANGELOG.md
Comment (Techassi) [inline crates/stackable-operator/CHANGELOG.md:22]: note: This needs to be removed before merging.

**Review by Techassi — CHANGES_REQUESTED** [review]
LGTM, only minor changes.

File: crates/stackable-operator/src/commons/s3/helpers.rs
Comment (NickLarsenNZ) [reply crates/stackable-operator/src/commons/s3/helpers.rs:111]: > This change should be operator-internal and not effect the user.

Sounds good to me.

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [inline crates/stackable-operator/src/commons/networking.rs:9]: note: This doc comments seems outdated, rewording needed.

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [inline crates/stackable-operator/src/commons/networking.rs:61]: note: Link to `IpAddr`.

```suggestion
/// A validated hostname, which is either a [`DomainName`] or [`IpAddr`].
```

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [inline crates/stackable-operator/src/commons/networking.rs:180]: note: Actually, no. This is not a valid domain name. Domain name labels **must** start with one of the 52 alphabetic characters (A-Z and a-z).

**Review by Techassi — CHANGES_REQUESTED** [review]
LGTM overall, just three small comments.

File: crates/stackable-operator/src/commons/networking.rs
Comment (sbernauer) [reply crates/stackable-operator/src/commons/networking.rs:61]: I'm always torn as this is our CRD description, and we should not leak Rust internals into our API. But yeah, anyway

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [reply crates/stackable-operator/src/commons/networking.rs:180]: See https://www.rfc-editor.org/rfc/rfc1035#section-2.3.1

File: crates/stackable-operator/src/commons/networking.rs
Comment (sbernauer) [reply crates/stackable-operator/src/commons/networking.rs:9]: Ohm I fail to see the outdated parts, can you please make a suggestion?

**Review by Techassi — APPROVED** [review]

File: crates/stackable-operator/src/commons/networking.rs
Comment (nightkr) [inline crates/stackable-operator/src/commons/networking.rs:76]: This is effectively dropping the validation regex from the composite type.

File: crates/stackable-operator/src/commons/networking.rs
Comment (nightkr) [inline crates/stackable-operator/src/commons/networking.rs:67]: Do we actually care about distinguishing these cases?

File: crates/stackable-operator/src/commons/networking.rs
Comment (nightkr) [inline crates/stackable-operator/src/commons/networking.rs:94]: This is dropping the context for _why_ validation failed.

File: crates/stackable-operator/src/commons/networking.rs
Comment (sbernauer) [reply crates/stackable-operator/src/commons/networking.rs:76]: You're right. So far we don't have any regex on HostName, but instead let IpAddr::parse do it's thing and try to parse it as domain name in case it#s not an IP.
We should be able to come up with an regex that accepts a DomainName, IPv4 or IPv6 address. But do you think it's worth the effort?

File: crates/stackable-operator/src/commons/networking.rs
Comment (sbernauer) [reply crates/stackable-operator/src/commons/networking.rs:67]: I personally modeled it as String, but @Techassi suggested using an enum instead, which sounded fine to me. No preference from me.

File: crates/stackable-operator/src/commons/networking.rs
Comment (nightkr) [reply crates/stackable-operator/src/commons/networking.rs:76]: That's fair.

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [reply crates/stackable-operator/src/commons/networking.rs:67]: The RFC calls for these two distinct variants of a hostname.

Why should we remove all the value we get from using proper types by just using a plain string? We _could_ simplify the `DomainName` variant to use `String`, but I would definitely keep the enum.

File: crates/stackable-operator/src/commons/networking.rs
Comment (nightkr) [reply crates/stackable-operator/src/commons/networking.rs:94]: That's one way, but it loses the error chain. Another alternative would be to guess based on a cursory inspection.. something like:

```rs
if value.starts_with(char::is_ascii_digit) || value.contains(':') {
    Ok(HostName::IpAddress(value.parse().context(InvalidIpAddressSnafu { hostname: value })?))
} else {
    Ok(HostName::DomainName(value.parse().context(InvalidDomainNameSnafu { hostname: value })?))
}
```

File: crates/stackable-operator/src/commons/networking.rs
Comment (nightkr) [reply crates/stackable-operator/src/commons/networking.rs:67]: Ultimately the question here is the purpose of the type. Detailed types makes sense for inspecting and manipulating the hostname. The original intent for `Hostname` was the much smaller scope of "make sure this opaque looks close enough to a hostname that it won't fuck up the syntax of the config file it's string-templated into".

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [reply crates/stackable-operator/src/commons/networking.rs:94]: We could also do something like:

```rust
match value.parse::<IpAddr>() {
    Ok(ip_addr) => return Ok(Hostname::IpAddr(ip_addr)),
    Err(ip_err) => match value.parse::<DomainName>() {
        Ok(dn) => return Ok(Hostname::DomainName(dn)),
        Err(dn_err) => // return both errors
    }
}
```

Comment (nightkr) [reply crates/stackable-operator/src/commons/networking.rs:94]: Yes that's effectively what @sbernauer suggested. There problem is that there's no way for an `Error` to have multiple `source()`s, so we'd have to pick one primary and render the other one as unawkwardly as we can.

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [reply crates/stackable-operator/src/commons/networking.rs:67]: > looks close enough

If we want to achieve this, sure. But then why reference RFC 1123 and only vaguely conform to it. If we want to conform to various RFCs, we should do so as closely as possible.

Another region where we _try_ to conform to an RFC (1035), we also completely miss the mark when validating domain labels. So I'm just trying to avoid the same thing happening here (even tho the domain name is not 100% valid).

File: crates/stackable-operator/src/commons/networking.rs
Comment (Techassi) [reply crates/stackable-operator/src/commons/networking.rs:94]: >  there's no way for an Error to have multiple source()s, so we'd have to pick one primary and render the other one as unawkwardly as we can.

That is indeed very true (and unfortunate). The cursed inspection is also kinda _meh_ tho. I don't know if it helps (I assume not), RFC 1123 states that the host name should first tried to be parsed as an IP address. So at least we do the correct thing RFC-wise.

**Pull request merged**
